Add doc for custom fields (#17)
IntuitDeveloper/intuit-api-docs: docs/use-cases/custom-fields.md

---
layout: default
title: Custom Fields
nav_order: 5
parent: Use Cases
---

## Custom Fields 

Custom fields could be business-specific fields that can be optionally configured. 
For QuickBooks Online Simple Start, Plus, and Essentials, 3 custom fields can be defined and set.

> For QuickBooks Advance, upto 12 custom fields can be configured.

Reviewer: ```suggestion
For QuickBooks Advance, up to 12 custom fields can be configured.
```

The API supports reading/writing their string representation.
This page describes how to configure/read these fields, and their values for transactional entities like Invoice Draft, Sales Receipt Draft, and Estimate Draft.
Along with that, this page outlines the API for reading all the defined custom fields for a company.

### Query for reading all fields defined for a company

```
query {
  customFields {
    id
    name
    inactive
    associatedEntityTypes {
      type 
      subtype
    }
    __typename
    ... on TextField {
      allowedValues {
        value
        id
        inactive
      }
    }
  }
 }
```
For dropdown type of custom fields, the list of allowedValues is defined so that when the field takes a value, it has to be one of the allowed set.

### Query for reading a field by ID

Similar filters can be applied to read based on name, inactive status

```
query {
  customFields (filter: {id: {equals: "CUSTOM_FIELD_ID"}}){
    id
    name
    inactive
    associatedEntityTypes {
      type 
      subtype
    }
    __typename
    ... on TextField {
      allowedValues {
        value
        id

>         inactive
>         
>       }

Reviewer: ```suggestion
        inactive
      }
```

    }
  }
 }
```

### Mutation for defining value of field within an Invoice Draft

**Note**  

> The following mutation shows just the fields required for creating the custom fields object within an Invoice Draft. For all the necessary field for creating an Invoice Draft, refer to the documentation for Invoice Draft.   

Reviewer: ```suggestion
The following mutation shows just the fields required for creating the custom fields object within an Invoice Draft. For all the necessary fields for creating an Invoice Draft, refer to the documentation for Invoice Draft.   
```

Reviewer: Also, perhaps we should add a link, where it says "refer to the documentation for Invoice Draft".  This could directly link to that page.

```
mutation createInvoiceDraft($invoiceDraftDetails: InvoiceDraftInput!) {
  createInvoiceDraft(invoiceDraftDetails: $invoiceDraftDetails) {
    id
    customFields { 
      value
      fieldId
      fieldName
    }
  }
}
```

### Variables
**Note**  

> The following JSON has just the fields required for custom fields object, for all the necessary variables for an Invoice Draft, refer to the documentation for Invoice Draft.

Reviewer: Do we need this note twice?  Maybe just once, either here or above.  I guess here makes sense, it's where the fields are defined.

### Variables
**Note**  
The following JSON has just the fields required for custom fields object.

> For all the necessary variables for an Invoice Draft, refer to the documentation for Invoice Draft.

Reviewer: Do we have any documentation for Invoice Draft yet?
